django/views/decorators/common.py

```diff
@@ -0,0 +1,14 @@
+from functools import wraps
+
+
+def no_append_slash(view_func):
+    """
+    Mark a view function as excluded from CommonMiddleware's APPEND_SLASH
+    redirection.
+    """
+    # view_func.should_append_slash = False would also work, but decorators are
+    # nicer if they don't have side effects, so return a new function.
```

Comment:
I'm not sure I really agree with this. The whole point of decorators is that they change the behaviour of the existing function. Wrapping the function here with another doesn't have the "side effect" of assigning a simple attribute to that original function, but then the original function is (strictly speaking) no longer in scope having been replaced by the wrapped function. Therefore it is unlikely anyone would use it anyway.

This is really just adding an extra level of indirection by wrapping in an additional function which then forces us to use `functools.wraps()` to fix things up. IMO the following is straightforward and achieves the same goal:

```python
def no_append_slash(func):
    func.should_append_slash = False
    return func
```

Reply:
Note that there are similar decorators that add attributes to functions in the works [here](https://github.com/django/django/pull/13483/files#diff-3fd16120fca44bdc94983057b9c28e7e815ed56a44f74fee21c2ddfac5cee2acR115-R139) and [here](https://github.com/django/django/pull/13532/files#diff-fd2291d1e25086ffa9e332f7bb4d16800501b7944a3659b1e5eb0a34e6ac924cR1-R60) that are not using this approach.

Reply:
Note that there are decorators that are using the same approach, e.g. `csrf_exempt()` :wink:

Reply:
Yes, but I have to ask why? I get that there is a side effect, but how much of a problem is this? I can potentially see the problem if you do the following:

```python
def view(request):
    ...

wrapped_view = no_append_slash(wrapped_view)

# Now both ``view`` and ``wrapped_view`` have ``should_append_slash`` set which
# means using ``view`` might not do what you expect.
```

But typically you're going to do the following:

```python
@no_append_slash
def view(request):
    ...
```

Which would be no different to assigning the attribute directly after the function is created if not using a wrapper function:

```python
def view(request):
    ...
view.should_append_slash = False
```

I was concerned that there may be an issue if you wanted to do something like the following, but it turns out that `functools.wraps()` also copies attributes other than `__name__` and `__doc__` so this works fine:

```python
@csrf_exempt
@no_append_slash
def view(request):
    ...
```

But it also effectively translates to having to call three functions, the outer two which have no benefit other than to have copies of all of the attributes of the functions within them. This has some call overhead and, if there are many decorated functions, excess memory use to store all these additional function objects and extra attributes as they bubble up.

Reply:
IMO supporting a typical usage is not enough in this case, because `no_append_slash` should be used to avoid enumerations, so it's security-related, like `csrf_exempt`. It must be bulletproof, without any side effects.

Reply:
I had a squint at this too, but didn't comment. But I agree with Nick - there isn't much of a reason to have an intermediate function, especially since decorators in Django aren't consistent. Mostly I see this as adding overhead and complicating debugging with an extra layer in stack traces.

Reply:
So here's the question, would you remove the same usage from `csrf_exempt`? (Would you take that risk?)

Reply:
I think the decorator strategy is orthogonal to this issue. Maybe there's a case for adjusting styles, or being consistent, but if so, we should address separately, where we can give it due consideration. 

2¢: Sometimes, sure, you don't use `wraps`, but I don't see it's addition as an issue. (TBH don't really buy any of the call overhead, memory use, or stack trace points…)